File: examples/yolov8_pose/cpp/yolov8_pose.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <cstdio>      // NOLINT(build/include_order)
#include <filesystem>  // NOLINT(build/c++17) NOLINT(build/include_order)
#include <iostream>    // NOLINT(build/include_order)
#include <memory>      // NOLINT(build/include_order)
#include <string>      // NOLINT(build/include_order)
#include <vector>      // NOLINT(build/include_order)

#include <opencv2/opencv.hpp>  // NOLINT(build/include_order)
#include <yaml-cpp/yaml.h>     // NOLINT(build/include_order)

#include "vision_service.h"  // NOLINT(build/include_order)

void print_usage(const char* program_name) {
    std::cout << "Usage: " << program_name << " <config_yaml> [options]\n"
                << "Options (any order after config_yaml):\n"
                << "  --model-path <path>   Override model_path in yaml\n"
                << "  --image <path>        Input image path (overrides config)\n"
                << "  --output <path>       Output image path (default: yolov8_pose_result.jpg)\n"
                << "  --use-camera          Use camera input instead of image file\n"
                << "  --camera-id <i>       Camera device ID (default: 0)\n"
                << "  --help                Show this help message\n"
                << "\nExample:\n"
                << "  " << program_name << " examples/yolov8_pose/config/yolov8_pose.yaml\n"
                << "  " << program_name
                << " examples/yolov8_pose/config/yolov8_pose.yaml --image test.jpg --output result.jpg\n";
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        print_usage(argv[0]);
        return 1;
    }

    std::string config_path = argv[1];
    std::string image_path;
    std::string output_path = "yolov8_pose_result.jpg";
    std::string model_path_override;
    bool use_camera = false;
    int camera_id = 0;

    for (int i = 2; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--help") {
            print_usage(argv[0]);
            return 0;
        } else if (arg == "--image" && i + 1 < argc) {
            image_path = argv[++i];
        } else if (arg == "--output" && i + 1 < argc) {
            output_path = argv[++i];
        } else if (arg == "--use-camera") {
            use_camera = true;
        } else if (arg == "--camera-id" && i + 1 < argc) {
            camera_id = std::stoi(argv[++i]);
        } else if (arg == "--model-path" && i + 1 < argc) {
            model_path_override = argv[++i];
        }
    }

    std::unique_ptr<VisionService> service = VisionService::Create(
        config_path,
        model_path_override,
        true);
    if (!service) {
        std::cerr << "Error: " << VisionService::LastCreateError() << std::endl;
        return 1;
    }

    if (use_camera) {
        std::cout << "Using camera " << camera_id << "..." << std::endl;
        cv::VideoCapture cap(camera_id);
        if (!cap.isOpened()) {
            std::cerr << "Error: Could not open camera " << camera_id << std::endl;
            return 1;
        }
        std::cout << "Real-time pose estimation. Press 'q' to quit, 's' to save..." << std::endl;
        cv::Mat frame;
        int frame_count = 0;
        double t_prev = static_cast<double>(cv::getTickCount()) / cv::getTickFrequency();
        double fps = 0.0;
        while (cap.read(frame)) {
            frame_count++;
            std::vector<VisionServiceResult> results;
            VisionServiceStatus ret = service->InferImage(frame, &results);
            if (ret != VISION_SERVICE_OK) {
                std::cerr << "Error: " << service->LastError() << std::endl;
                cap.release();
                cv::destroyAllWindows();
                return 1;
            }
            cv::Mat vis;
            if (!results.empty()) {
                if (frame_count <= 5 || frame_count % 30 == 0)
                    std::cout << "Frame " << frame_count << ": Detected " << results.size() << " persons" << std::endl;
                service->Draw(frame, &vis);
            } else {
                vis = frame;
            }
            char fps_buf[32];
            std::snprintf(fps_buf, sizeof(fps_buf), "FPS: %.1f", fps);
            cv::putText(vis, fps_buf, cv::Point(10, 30), cv::FONT_HERSHEY_SIMPLEX, 1, cv::Scalar(0, 255, 0), 2);
            cv::imshow("YOLOv8 Pose", vis);
            char key = cv::waitKey(1) & 0xFF;
            if (key == 'q') break;
            if (key == 's') {
                std::string save_path = "camera_frame_" + std::to_string(frame_count) + ".jpg";
                cv::imwrite(save_path, vis);
                std::cout << "Saved frame to: " << save_path << std::endl;
            }
            double t_now = static_cast<double>(cv::getTickCount()) / cv::getTickFrequency();
            fps = (t_now - t_prev) > 1e-6 ? 1.0 / (t_now - t_prev) : 0.0;
            t_prev = t_now;
        }
        cap.release();
        cv::destroyAllWindows();
        std::cout << "Processed " << frame_count << " frames." << std::endl;
    } else {
        if (image_path.empty()) {
            const std::string default_image = service->GetDefaultImage();
            if (!default_image.empty()) {
                image_path = default_image;
            }
        }
        if (image_path.empty()) {
            std::cerr << "Error: No input image. Use --image <path> or set test_image in config." << std::endl;
            return 1;
        }
        std::cout << "Running pose estimation..." << std::endl;
        cv::Mat img = cv::imread(image_path);
        if (img.empty()) {
            std::cerr << "Error: Could not load image: " << image_path << std::endl;
            return 1;
        }
        std::vector<VisionServiceResult> results;
        VisionServiceStatus ret = service->InferImage(img, &results);
        if (ret != VISION_SERVICE_OK) {
            std::cerr << "Error: " << service->LastError() << std::endl;
            return 1;
        }
        if (!results.empty()) {
            std::cout << "Detected " << results.size() << " persons with keypoints" << std::endl;
            cv::Mat vis;
            service->Draw(img, &vis);
            cv::imwrite(output_path, vis);
            std::cout << "Result image saved to: " << output_path << std::endl;
        } else {
            std::cout << "No persons detected" << std::endl;
        }
    }
    return 0;
}

File: examples/yolov8_pose/python/yolov8_pose.py
#!/usr/bin/env python3
# Copyright 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
#
# SPDX-License-Identifier: Apache-2.0

"""
YOLOv8 Pose Estimation Example using CV Model Factory

运行方式（与其它 examples 一致）：
  1. 通过 --config 指定 yaml 路径
  2. 通过 VisionServiceNative.create(config_path) 创建模型
  3. 从 yaml 读取参数（test_image 等）进行推理
"""

import argparse
from pathlib import Path
import time
import cv2
import yaml

from spacemit_vision import VisionServiceNative, VisionServiceStatus


def resolve_path(path_value, project_root):
    p = Path(path_value).expanduser()
    return p if p.is_absolute() else (project_root / p).resolve()


def parse_args():
    parser = argparse.ArgumentParser(description="YOLOv8 Pose Estimation Example")
    parser.add_argument("--config", type=str, default=None,
                       help="Config yaml path (default: examples/yolov8_pose/config/yolov8_pose.yaml)")
    parser.add_argument("--model-path", type=str, default=None,
                       help="Override model_path in yaml")
    parser.add_argument("--image", type=str,
                       help="Input image path (if not provided, uses config default)")
    parser.add_argument("--output", type=str, default="yolov8_pose_result.jpg",
                       help="Output image path")
    parser.add_argument("--conf-threshold", type=float, default=None,
                       help="Confidence threshold (default: from config yaml)")
    parser.add_argument("--iou-threshold", type=float, default=None,
                       help="IoU threshold for NMS (default: from config yaml)")
    parser.add_argument("--use-camera", action="store_true",
                       help="Use camera input instead of image file")
    parser.add_argument("--camera-id", type=int, default=0,
                       help="Camera device ID (default: 0)")
    return parser.parse_args()


def main():
    args = parse_args()

    try:
        # 1) 本 example 的 config 目录；资源路径（test_image 等）相对 model_zoo/cv
        default_config = Path(__file__).parent.parent / "config" / "yolov8_pose.yaml"
        config_path = Path(args.config) if args.config else default_config
        project_root = Path(__file__).parent.parent.parent.parent  # model_zoo/cv

        model_name = config_path.stem
        # 2) 根据 yaml 创建模型
        print(f"创建 {model_name} 姿态估计器...")
        estimator = VisionServiceNative.create(
            str(config_path), model_path_override=args.model_path or ""
        )

        # 3) 从同一 yaml 读取推理参数（test_image 等）
        config = {}
        label_file = config_path
        if label_file.exists():
            with open(label_file, "r", encoding="utf-8") as f:
                config = yaml.safe_load(f) or {}

        # Handle camera or image input
        if args.use_camera:
            print(f"使用摄像头 {args.camera_id}...")
            cap = cv2.VideoCapture(args.camera_id)
            if not cap.isOpened():
                raise ValueError(f"无法打开摄像头 {args.camera_id}")

            print("实时姿态估计中，按 'q' 退出，按 's' 保存当前帧...")
            frame_count = 0
            t_prev = time.perf_counter()
            fps = 0.0

            while True:
                ret, frame = cap.read()
                if not ret:
                    print("无法读取摄像头帧")
                    break

                frame_count += 1

                # Run pose estimation
                status, results = estimator.infer_image(
                    frame,
                    conf=args.conf_threshold if args.conf_threshold is not None else -1.0,
                    iou=args.iou_threshold if args.iou_threshold is not None else -1.0,
                )
                if status != VisionServiceStatus.OK:
                    raise RuntimeError(estimator.last_error())

                if results:
                    # Draw results using C++ side drawing API
                    if estimator.supports_draw():
                        st, out = estimator.draw(frame)
                        result_frame = out if st == VisionServiceStatus.OK else frame.copy()
                    else:
                        result_frame = frame.copy()

                    # Print detection info for first few frames
                    if frame_count <= 5 or frame_count % 30 == 0:
                        print(f"帧 {frame_count}: 检测到 {len(results)} 个人")
                else:
                    result_frame = frame.copy()
                    if frame_count <= 5 or frame_count % 30 == 0:
                        print(f"帧 {frame_count}: 未检测到人")

                cv2.putText(result_frame, f"FPS: {fps:.1f}", (10, 30),
                            cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)
                cv2.imshow("YOLOv8 Pose Estimation", result_frame)
                key = cv2.waitKey(1) & 0xFF
                if key == ord('q'):
                    break
                elif key == ord('s'):
                    save_path = f"camera_frame_{frame_count}.jpg"
                    cv2.imwrite(save_path, result_frame)
                    print(f"保存帧到: {save_path}")

                t_now = time.perf_counter()
                fps = 1.0 / (t_now - t_prev) if (t_now - t_prev) > 1e-6 else 0.0
                t_prev = t_now

            cap.release()
            cv2.destroyAllWindows()
            print(f"已处理 {frame_count} 帧")
        else:
            # 输入图像：命令行优先，否则用 yaml 中的 test_image
            if args.image:
                image_path = resolve_path(args.image, project_root)
            else:
                image_path = config.get("test_image", "~/.cache/assets/image/013_pose.jpg")
                image_path = resolve_path(image_path, project_root)

            # Load image
            print(f"加载图像: {image_path}")
            image = cv2.imread(str(image_path))
            if image is None:
                raise FileNotFoundError(f"无法加载图像: {image_path}")

            print(f"图像尺寸: {image.shape}")

            # Run pose estimation
            print("运行姿态估计...")
            status, results = estimator.infer_image(
                image,
                conf=args.conf_threshold if args.conf_threshold is not None else -1.0,
                iou=args.iou_threshold if args.iou_threshold is not None else -1.0,
            )
            if status != VisionServiceStatus.OK:
                raise RuntimeError(estimator.last_error())

            # Process results
            print(f"检测到 {len(results)} 个人:")

            for i, r in enumerate(results):
                x1, y1, x2, y2 = int(r.x1), int(r.y1), int(r.x2), int(r.y2)
                confidence = r.score

                print(f"  人 {i+1}: 置信度 {confidence:.3f} at [{x1}, {y1}, {x2}, {y2}]")

                # Get keypoints (each has .x .y .visibility)
                keypoints = r.keypoints
                print(f"    关键点数量: {len(keypoints)}")

                # Count visible keypoints
                visible_count = sum(1 for kpt in keypoints if kpt.visibility > 0.5)
                print(f"    可见关键点: {visible_count}/{len(keypoints)}")

            # Draw results using C++ side drawing API
            if estimator.supports_draw():
                st, out = estimator.draw(image)
                result_image = out if st == VisionServiceStatus.OK else image
            else:
                result_image = image

            # Save result
            cv2.imwrite(args.output, result_image)
            print(f"结果已保存到: {args.output}")


    except Exception as e:
        print(f"错误: {e}")
        return 1

    return 0


if __name__ == "__main__":
    exit(main())
